File: samples/csharp/image-analysis/dotnetcore/Samples.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
// Azure AI Vision SDK -- C# Image Analysis Samples
//
using Azure;
using Azure.AI.Vision.Core.Input;
using Azure.AI.Vision.Core.Options;
using Azure.AI.Vision.ImageAnalysis;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ImageAnalysisSamples
{
    public class Samples
    {
        // This sample does analysis on an image file using all visual features, and prints the results to the console,
        // including the detailed results.
        public static void GetAllResults(string endpoint, string key)
        {
            var serviceOptions = new VisionServiceOptions(endpoint, new AzureKeyCredential(key));

            // Specify the image file on disk to analyze. sample1.jpg is a good example to show most features,
            // except Text (OCR). Use sample2.jpg for OCR.
            using var imageSource = VisionSource.FromFile("sample1.jpg");

            // Or, instead of the above, specify a publicly accessible image URL to analyze
            // (e.g. https://learn.microsoft.com/azure/cognitive-services/computer-vision/images/windows-kitchen.jpg)
            // var imageSource = VisionSource.FromUrl(new Uri("YourImageURL"));

            var analysisOptions = new ImageAnalysisOptions()
            {
                // Mandatory. You must set one or more features to analyze. Here we use the full set of features.
                // Note that 'Caption' and 'DenseCaptions' are only supported in Azure GPU regions (East US, France Central, Korea Central,
                // North Europe, Southeast Asia, West Europe, West US). Remove 'Caption' and 'DenseCaptions' from the list below if your
                // Computer Vision key is not from one of those regions.
                Features =
                      ImageAnalysisFeature.CropSuggestions
                    | ImageAnalysisFeature.Caption
                    | ImageAnalysisFeature.DenseCaptions
                    | ImageAnalysisFeature.Objects
                    | ImageAnalysisFeature.People
                    | ImageAnalysisFeature.Text
                    | ImageAnalysisFeature.Tags,

                // Optional, and only relevant when you select ImageAnalysisFeature.CropSuggestions.
                // Define one or more aspect ratios for the desired cropping. Each aspect ratio needs to be in the range [0.75, 1.8].
                // If you do not set this, the service will return one crop suggestion with the aspect ratio it sees fit.
                CroppingAspectRatios = new List<double>() { 0.9, 1.33 },

                // Optional. Default is "en" for English. See https://aka.ms/cv-languages for a list of supported
                // language codes and which visual features are supported for each language.
                Language = "en",

                // Optional. Default is "latest".
                ModelVersion = "latest",

                // Optional, and only relevant when you select ImageAnalysisFeature.Caption.
                // Set this to "true" to get a gender neutral caption (the default is "false").
                GenderNeutralCaption = true
            };

            using var analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions);

            Console.WriteLine(" Please wait for image analysis results...\n");

            // This call creates the network connection and blocks until Image Analysis results
            // return (or an error occurred). Note that there is also an asynchronous (non-blocking)
            // version of this method: analyzer.AnalyzeAsync().
            var result = analyzer.Analyze();

            if (result.Reason == ImageAnalysisResultReason.Analyzed)
            {
                Console.WriteLine($" Image height = {result.ImageHeight}");
                Console.WriteLine($" Image width = {result.ImageWidth}");
                Console.WriteLine($" Model version = {result.ModelVersion}");

                if (result.Caption != null)
                {
                    Console.WriteLine(" Caption:");
                    Console.WriteLine($"   \"{result.Caption.Content}\", Confidence {result.Caption.Confidence:0.0000}");
                }

                if (result.DenseCaptions != null)
                {
                    Console.WriteLine(" Dense Captions:");
                    foreach (var caption in result.DenseCaptions)
                    {
                        Console.WriteLine($"   \"{caption.Content}\", Bounding box {caption.BoundingBox}, Confidence {caption.Confidence:0.0000}");
                    }
                }

                if (result.Objects != null)
                {
                    Console.WriteLine(" Objects:");
                    foreach (var detectedObject in result.Objects)
                    {
                        Console.WriteLine($"   \"{detectedObject.Name}\", Bounding box {detectedObject.BoundingBox}, Confidence {detectedObject.Confidence:0.0000}");
                    }
                }

                if (result.Tags != null)
                {
                    Console.WriteLine($" Tags:");
                    foreach (var tag in result.Tags)
                    {
                        Console.WriteLine($"   \"{tag.Name}\", Confidence {tag.Confidence:0.0000}");
                    }
                }

                if (result.People != null)
                {
                    Console.WriteLine($" People:");
                    foreach (var person in result.People)
                    {
                        Console.WriteLine($"   Bounding box {person.BoundingBox}, Confidence {person.Confidence:0.0000}");
                    }
                }

                if (result.CropSuggestions != null)
                {
                    Console.WriteLine($" Crop Suggestions:");
                    foreach (var cropSuggestion in result.CropSuggestions)
                    {
                        Console.WriteLine($"   Aspect ratio {cropSuggestion.AspectRatio}: "
                            + $"Crop suggestion {cropSuggestion.BoundingBox}");
                    };
                }

                if (result.Text != null)
                {
                    Console.WriteLine($" Text:");
                    foreach (var line in result.Text.Lines)
                    {
                        string pointsToString = "{" + string.Join(',', line.BoundingPolygon.Select(pointsToString => pointsToString.ToString())) + "}";
                        Console.WriteLine($"   Line: '{line.Content}', Bounding polygon {pointsToString}");

                        foreach (var word in line.Words)
                        {
                            pointsToString = "{" + string.Join(',', word.BoundingPolygon.Select(pointsToString => pointsToString.ToString())) + "}";
                            Console.WriteLine($"     Word: '{word.Content}', Bounding polygon {pointsToString}, Confidence {word.Confidence:0.0000}");
                        }
                    }
                }

                var resultDetails = ImageAnalysisResultDetails.FromResult(result);
                Console.WriteLine($" Result details:");
                Console.WriteLine($"   Image ID = {resultDetails.ImageId}");
                Console.WriteLine($"   Result ID = {resultDetails.ResultId}");
                Console.WriteLine($"   Connection URL = {resultDetails.ConnectionUrl}");
                Console.WriteLine($"   JSON result = {resultDetails.JsonResult}");
            }
            else if (result.Reason == ImageAnalysisResultReason.Error)
            {
                var errorDetails = ImageAnalysisErrorDetails.FromResult(result);
                Console.WriteLine(" Analysis failed.");
                Console.WriteLine($"   Error reason : {errorDetails.Reason}");
                Console.WriteLine($"   Error code : {errorDetails.ErrorCode}");
                Console.WriteLine($"   Error message: {errorDetails.Message}");
                Console.WriteLine(" Did you set the computer vision endpoint and key?");
            }
        }

        // This sample does analysis on an image URL, showing how to use the Analyzed event to get
        // the analysis result for one visual feature (tags).
        public static async Task GetResultsUsingAnalyzedEvent(string endpoint, string key)
        {
            var serviceOptions = new VisionServiceOptions(endpoint, new AzureKeyCredential(key));

            using var imageSource = VisionSource.FromUrl(new Uri("https://docs.microsoft.com/azure/cognitive-services/computer-vision/images/windows-kitchen.jpg"));

            var analysisOptions = new ImageAnalysisOptions() { Features = ImageAnalysisFeature.Tags };

            using var analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions);

            var tcsEventReceived  = new TaskCompletionSource<bool>();

            analyzer.Analyzed += (_, e) =>
            {
                if (e.Result.Reason == ImageAnalysisResultReason.Analyzed)
                {
                    Console.WriteLine($" Tags:");
                    foreach (var tag in e.Result.Tags)
                    {
                        Console.WriteLine($"   \"{tag.Name}\", Confidence {tag.Confidence:0.0000}");
                    }
                }
                else if (e.Result.Reason == ImageAnalysisResultReason.Error)
                {
                    Console.WriteLine(" Analysis failed.");

                    var errorDetails = ImageAnalysisErrorDetails.FromResult(e.Result);
                    Console.WriteLine($"   Error reason : {errorDetails.Reason}");
                    Console.WriteLine($"   Error code : {errorDetails.ErrorCode}");
                    Console.WriteLine($"   Error message: {errorDetails.Message}");
                    Console.WriteLine(" Did you set the computer vision endpoint and key?");
                }

                tcsEventReceived.SetResult(true);
            };

            Console.WriteLine(" Please wait for image analysis results...");
            await analyzer.AnalyzeAsync();

            // Make sure we received the Analyzed event before exiting this method
            Task.WaitAny(tcsEventReceived.Task);
        }

        // This sample does analysis on an image file using a given custom-trained model, and shows how
        // to get the detected objects and/or tags.
        public static void GetCustomModelResults(string endpoint, string key)
        {
            var serviceOptions = new VisionServiceOptions(endpoint, new AzureKeyCredential(key));

            using var imageSource = VisionSource.FromFile("sample1.jpg");

            var analysisOptions = new ImageAnalysisOptions()
            {
                // Set your custom model name here
                ModelName = "MyCustomModelName"
            };

            using var analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions);

            var result = analyzer.Analyze();

            if (result.Reason == ImageAnalysisResultReason.Analyzed)
            {
                // Use the CustomObjects & CustomTags properties to get the results

                if (result.CustomObjects != null)
                {
                    Console.WriteLine(" Custom Objects:");
                    foreach (var detectedObject in result.CustomObjects)
                    {
                        Console.WriteLine($"   \"{detectedObject.Name}\", Bounding box {detectedObject.BoundingBox}, Confidence {detectedObject.Confidence:0.0000}");
                    }
                }

                if (result.CustomTags != null)
                {
                    Console.WriteLine($" Custom Tags:");
                    foreach (var tag in result.CustomTags)
                    {
                        Console.WriteLine($"   \"{tag.Name}\", Confidence {tag.Confidence:0.0000}");
                    }
                }
            }
            else if (result.Reason == ImageAnalysisResultReason.Error)
            {
                var errorDetails = ImageAnalysisErrorDetails.FromResult(result);
                Console.WriteLine(" Analysis failed.");
                Console.WriteLine($"   Error reason : {errorDetails.Reason}");
                Console.WriteLine($"   Error code : {errorDetails.ErrorCode}");
                Console.WriteLine($"   Error message: {errorDetails.Message}");
                Console.WriteLine(" Did you set the computer vision endpoint and key?");
            }
        }
    }
}

File: samples/java/image-analysis/src/azure/ai/vision/imageanalysis/samples/Samples.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See https://aka.ms/csspeech/license for the full license information.
//
package azure.ai.vision.imageanalysis.samples;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.IOException;
import java.net.URL;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.EnumSet;

import reactor.core.publisher.Mono;

import com.azure.ai.vision.common.*;
import com.azure.ai.vision.imageanalysis.*;

/**
 * This sample application shows how to use features of the Azure AI Vision
 * Image Analysis SDK for Java
 */
public class Samples {

    /**
     * This sample does analysis on an image file using all visual features
     * and a synchronous (blocking) call. It prints the results to the console,
     * including the detailed results.
     * 
     * @param endpoint The Azure endpoint for the Computer Vision service
     * @param key      The Azure key for the Computer Vision service
     */
    public static void analyzeImageFromFile(String endpoint, String key) {
        try (
            VisionServiceOptions serviceOptions = new VisionServiceOptions(new URL(endpoint), key);

            // Specify the image file on disk to analyze. sample.jpg is a good example to show most features.
            // Alternatively, specify an image URL (e.g. https://aka.ms/azai/vision/image-analysis-sample.jpg)
            // or a memory buffer containing the image. see:
            // https://learn.microsoft.com/azure/ai-services/computer-vision/how-to/call-analyze-image-40?pivots=programming-language-java#select-the-image-to-analyze
            VisionSource imageSource = VisionSource.fromFile("sample.jpg");

            ImageAnalysisOptions analysisOptions = new ImageAnalysisOptions()) {

            // Mandatory. You must set one or more features to analyze. Here we use the full set of features.
            // Note that 'Caption' and 'DenseCaptions' are only supported in Azure GPU regions (East US, France Central, Korea Central,
            // North Europe, Southeast Asia, West Europe, West US). Remove 'Caption' and 'DenseCaptions' from the list below if your
            // Computer Vision key is not from one of those regions.
            analysisOptions.setFeatures(EnumSet.of(
                    ImageAnalysisFeature.CROP_SUGGESTIONS,
                    ImageAnalysisFeature.CAPTION,
                    ImageAnalysisFeature.DENSE_CAPTIONS,
                    ImageAnalysisFeature.OBJECTS,
                    ImageAnalysisFeature.PEOPLE,
                    ImageAnalysisFeature.TEXT,
                    ImageAnalysisFeature.TAGS));
            
            // Optional, and only relevant when you select ImageAnalysisFeature.CropSuggestions.
            // Define one or more aspect ratios for the desired cropping. Each aspect ratio needs to be in the range [0.75, 1.8].
            // If you do not set this, the service will return one crop suggestion with the aspect ratio it sees fit.
            analysisOptions.setCroppingAspectRatios(Arrays.asList(0.9, 1.33));
            
            // Optional. Default is "en" for English. See https://aka.ms/cv-languages for a list of supported
            // language codes and which visual features are supported for each language.
            analysisOptions.setLanguage("en");
            
            // Optional. Default is "latest".
            analysisOptions.setModelVersion("latest");
            
            // Optional, and only relevant when you select ImageAnalysisFeature.Caption.
            // Set this to "true" to get a gender neutral caption (the default is "false").
            analysisOptions.setGenderNeutralCaption(true);

            try (
                ImageAnalyzer analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions)) {
                System.out.println(" Please wait for image analysis results...\n");

                // This call creates the network connection and blocks until Image Analysis results
                // return (or an error occurred). Note that there is also an asynchronous (non-blocking)
                // version of this method: analyzer.analyzeAsync().
                try(
                    ImageAnalysisResult result = analyzer.analyze()) {

                    if (result.getReason() == ImageAnalysisResultReason.ANALYZED) {
                        System.out.println(" Image height = " + result.getImageHeight());
                        System.out.println(" Image width = " + result.getImageWidth());
                        System.out.println(" Model version = " + result.getModelVersion());

                        if (result.getCaption() != null) {
                            System.out.println(" Caption:");
                            System.out.println("   \"" + result.getCaption().getContent() + "\", Confidence "
                                    + String.format("%.4f", result.getCaption().getConfidence()));
                        }

                        if (result.getDenseCaptions() != null) {
                            System.out.println(" Dense Captions:");
                            for (ContentCaption denseCaption : result.getDenseCaptions()) {
                                System.out.println("   \"" + denseCaption.getContent() + "\", Bounding box "
                                        + denseCaption.getBoundingBox() +
                                        ", Confidence " + String.format("%.4f", denseCaption.getConfidence()));
                            }
                        }

                        if (result.getObjects() != null) {
                            System.out.println(" Objects:");
                            for (DetectedObject detectedObject : result.getObjects()) {
                                System.out.println("   \"" + detectedObject.getName() + "\", Bounding box "
                                        + detectedObject.getBoundingBox() +
                                        ", Confidence " + String.format("%.4f", detectedObject.getConfidence()));
                            }
                        }

                        if (result.getTags() != null) {
                            System.out.println(" Tags:");
                            for (ContentTag tag : result.getTags()) {
                                System.out.println("   \"" + tag.getName() + "\", Confidence "
                                        + String.format("%.4f", tag.getConfidence()));
                            }
                        }

                        if (result.getPeople() != null) {
                            System.out.println(" People:");
                            for (DetectedPerson person : result.getPeople()) {
                                System.out.println("   Bounding box " + person.getBoundingBox() +
                                        ", Confidence " + String.format("%.4f", person.getConfidence()));
                            }
                        }

                        if (result.getCropSuggestions() != null) {
                            System.out.println(" Crop Suggestions:");
                            for (CropSuggestion cropSuggestion : result.getCropSuggestions()) {
                                System.out.println("   Aspect ratio " + cropSuggestion.getAspectRatio() +
                                        ": Crop suggestion " + cropSuggestion.getBoundingBox());
                            }
                        }

                        if (result.getText() != null) {
                            System.out.println(" Text:");
                            for (DetectedTextLine line : result.getText()) {
                                String pointsToString = "{" + String.join(",",
                                        line.getBoundingPolygon().stream().map(Object::toString).toArray(String[]::new))
                                        + "}";
                                System.out.println(
                                        "   Line: '" + line.getContent() + "', Bounding polygon " + pointsToString);

                                for (DetectedTextWord word : line.getWords()) {
                                    pointsToString = "{" + String.join(",",
                                            word.getBoundingPolygon().stream().map(Object::toString).toArray(String[]::new))
                                            + "}";
                                    System.out.println(
                                            "     Word: '" + word.getContent() + "', Bounding polygon " + pointsToString +
                                                    ", Confidence " + String.format("%.4f", word.getConfidence()));
                                }
                            }
                        }

                        ImageAnalysisResultDetails resultDetails = ImageAnalysisResultDetails.fromResult(result);
                        System.out.println(" Result details:");
                        System.out.println("   Image ID = " + resultDetails.getImageId());
                        System.out.println("   Result ID = " + resultDetails.getResultId());
                        System.out.println("   Connection URL = " + resultDetails.getConnectionUrl());
                        System.out.println("   JSON result = " + resultDetails.getJsonResult());
                    } else { // result.Reason == ImageAnalysisResultReason.Error
                        ImageAnalysisErrorDetails errorDetails = ImageAnalysisErrorDetails.fromResult(result);
                        System.out.println(" Analysis failed.");
                        System.out.println("   Error reason: " + errorDetails.getReason());
                        System.out.println("   Error code: " + errorDetails.getErrorCode());
                        System.out.println("   Error message: " + errorDetails.getMessage());
                        System.out.println(" Did you set the computer vision endpoint and key?");
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * This sample does analysis on an image URL, using an asynchronous
     * (non-blocking) call to analyze one visual feature (Tags)
     * 
     * @param endpoint The Azure endpoint for the Computer Vision service
     * @param key      The Azure key for the Computer Vision service
     */
    public static void analyzeImageFromUrlAsync(String endpoint, String key) {
        try (
            VisionServiceOptions serviceOptions = new VisionServiceOptions(new URL(endpoint), key);

            VisionSource imageSource = VisionSource.fromUrl(
                new URL("https://aka.ms/azai/vision/image-analysis-sample.jpg"));

            ImageAnalysisOptions analysisOptions = new ImageAnalysisOptions();) {

            analysisOptions.setFeatures(EnumSet.of(ImageAnalysisFeature.TAGS));

            try (
                ImageAnalyzer analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions)) {

                Mono<ImageAnalysisResult> analysisResultMono = analyzer.analyzeAsync();

                analysisResultMono.subscribe(result -> {
                    if (result.getReason() == ImageAnalysisResultReason.ANALYZED) {
                        System.out.println(" Tags:");
                        for (ContentTag tag : result.getTags()) {
                            System.out.println(
                                    String.format("   \"%s\", Confidence %.4f", tag.getName(), tag.getConfidence()));
                        }
                    } else {
                        System.out.println(" Analysis failed.");
                        ImageAnalysisErrorDetails errorDetails = ImageAnalysisErrorDetails.fromResult(result);
                        System.out.println("   Error reason: " + errorDetails.getReason());
                        System.out.println("   Error code: " + errorDetails.getErrorCode());
                        System.out.println("   Error message: " + errorDetails.getMessage());
                        System.out.println(" Did you set the computer vision endpoint and key?");
                    }
                });

                analysisResultMono.block().close();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * This sample does analysis on an image from an input memory buffer, using synchronous (blocking)
     * call to analyze one visual feature (Caption).
     * 
     * @param endpoint The Azure endpoint for the Computer Vision service
     * @param key      The Azure key for the Computer Vision service
     */
    public static void analyzeImageFromBuffer(String endpoint, String key) {

        try {
            // This sample assumes you have an image in a memory buffer. Here we simply load it from file.
            final ByteBuffer imageBuffer = fileToByteBuffer("sample.jpg");

            try (
                // Create an ImageSourceBuffer, and copy the input image into it
                ImageSourceBuffer imageSourceBuffer = new ImageSourceBuffer();
                ImageWriter imageWriter = imageSourceBuffer.getWriter()) {
                imageWriter.write(imageBuffer);

                try (
                    // Create your VisionSource from the ImageSourceBuffer
                    VisionSource imageSource = VisionSource.fromImageSourceBuffer(imageSourceBuffer);

                    ImageAnalysisOptions analysisOptions = new ImageAnalysisOptions()) {
                    analysisOptions.setFeatures(EnumSet.of(ImageAnalysisFeature.CAPTION));

                    try (
                        VisionServiceOptions serviceOptions = new VisionServiceOptions(new URL(endpoint), key);
                        ImageAnalyzer analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions);
                        ImageAnalysisResult result = analyzer.analyze()) {

                        if (result.getReason() == ImageAnalysisResultReason.ANALYZED) {

                            if (result.getCaption() != null) {
                                System.out.println(" Caption:");
                                System.out.println("   \"" + result.getCaption().getContent() + "\", Confidence "
                                        + String.format("%.4f", result.getCaption().getConfidence()));
                            }

                        } else { // result.Reason == ImageAnalysisResultReason.Error

                            ImageAnalysisErrorDetails errorDetails = ImageAnalysisErrorDetails.fromResult(result);
                            System.out.println(" Analysis failed.");
                            System.out.println("   Error reason: " + errorDetails.getReason());
                            System.out.println("   Error code: " + errorDetails.getErrorCode());
                            System.out.println("   Error message: " + errorDetails.getMessage());
                            System.out.println(" Did you set the computer vision endpoint and key?");
                        }
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * This sample does analysis on an image file using a given custom-trained
     * model, and shows how to get the detected objects and/or tags.
     * 
     * @param endpoint The Azure endpoint for the Computer Vision service
     * @param key      The Azure key for the Computer Vision service
     */
    public static void analyzeImageWithCustomModel(String endpoint, String key) {
        try (
            VisionServiceOptions serviceOptions = new VisionServiceOptions(new URL(endpoint), key);

            VisionSource visionSource = VisionSource.fromFile("sample.jpg");

            ImageAnalysisOptions analysisOptions = new ImageAnalysisOptions();) {

            // Set the model name to the name of your custom model
            analysisOptions.setModelName("YourCustomModelNameHere");

            try (
                ImageAnalyzer analyzer = new ImageAnalyzer(serviceOptions, visionSource, analysisOptions);

                ImageAnalysisResult result = analyzer.analyze()) {

                if (result.getReason() == ImageAnalysisResultReason.ANALYZED) {

                    if (result.getCustomObjects() != null) {
                        System.out.println("Custom Objects:");
                        for (DetectedObject detectedObject : result.getCustomObjects()) {
                            System.out.println(String.format("\"%s\", Bounding box %s, Confidence %.4f",
                                    detectedObject.getName(), detectedObject.getBoundingBox(),
                                    detectedObject.getConfidence()));
                        }
                    }

                    if (result.getCustomTags() != null) {
                        System.out.println("Custom Tags:");
                        for (ContentTag tag : result.getCustomTags()) {
                            System.out.println(String.format("   \"%s\", Confidence %.4f",
                                    tag.getName(), tag.getConfidence()));
                        }
                    }

                } else { // result.getReason() == ImageAnalysisResultReason.ERROR
                    ImageAnalysisErrorDetails errorDetails = ImageAnalysisErrorDetails.fromResult(result);
                    System.out.println("Analysis failed.");
                    System.out.println("Error reason: " + errorDetails.getReason());
                    System.out.println("Error code: " + errorDetails.getErrorCode());
                    System.out.println("Error message: " + errorDetails.getMessage());
                    System.out.println("Did you set the computer vision endpoint and key?");
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * This sample does segmentation of an input image and writes the
     * resulting background-removed image or foreground matte image to disk.
     * 
     * @param endpoint The Azure endpoint for the Computer Vision service
     * @param key      The Azure key for the Computer Vision service
     */
    public static void backgroundRemoval(String endpoint, String key) {
        try (
            VisionServiceOptions serviceOptions = new VisionServiceOptions(new URL(endpoint), key);

            // Specify the image file on disk to analyze. sample.jpg is a good example to show most features.
            // Alternatively, specify an image URL (e.g. https://aka.ms/azai/vision/image-analysis-sample.jpg)
            // or a memory buffer containing the image. see:
            // https://learn.microsoft.com/azure/ai-services/computer-vision/how-to/call-analyze-image-40?pivots=programming-language-java#select-the-image-to-analyze
            VisionSource imageSource = VisionSource.fromFile("sample.jpg");

            ImageAnalysisOptions analysisOptions = new ImageAnalysisOptions()) {

            // Set one of two segmentation options: 'ImageSegmentationMode.BACKGROUND_REMOVAL' or
            // 'ImageSegmentationMode.FOREGROUND_MATTING'
            analysisOptions.setSegmentationMode(ImageSegmentationMode.BACKGROUND_REMOVAL);

            try (
                ImageAnalyzer analyzer = new ImageAnalyzer(serviceOptions, imageSource, analysisOptions)) {

                System.out.println(" Please wait for image analysis results...\n");

                // This call creates the network connection and blocks until Image Analysis
                // results return (or an error occurred). Note that there is also an
                // asynchronous
                // (non-blocking) version of this method: analyzer.analyzeAsync().
                try (
                    ImageAnalysisResult result = analyzer.analyze()) {

                    if (result.getReason() == ImageAnalysisResultReason.ANALYZED) {

                        SegmentationResult segmentationResult = result.getSegmentationResult();

                        // Get the resulting output image buffer (PNG format)
                        byte[] imageBuffer = segmentationResult.getImageBuffer();
                        System.out.println(" Segmentation result:");
                        System.out.println("   Output image buffer size (bytes) = " + imageBuffer.length);

                        // Get output image size
                        System.out.println("   Output image height = " + segmentationResult.getImageHeight());
                        System.out.println("   Output image width = " + segmentationResult.getImageWidth());

                        // Write the buffer to a file
                        String outputImageFile = "output.png";
                        try (FileOutputStream fos = new FileOutputStream(outputImageFile)) {
                            fos.write(imageBuffer);
                        }
                        System.out.println("   File " + outputImageFile + " written to disk");

                    } else { // result.getReason() == ImageAnalysisResultReason.ERROR
                        ImageAnalysisErrorDetails errorDetails = ImageAnalysisErrorDetails.fromResult(result);
                        System.out.println(" Analysis failed.");
                        System.out.println("   Error reason: " + errorDetails.getReason());
                        System.out.println("   Error code: " + errorDetails.getErrorCode());
                        System.out.println("   Error message: " + errorDetails.getMessage());
                        System.out.println(" Did you set the computer vision endpoint and key?");
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Helper method to return a ByteBuffer containing content
     * read from file on disk.
     * 
     * @param path The full-path file name of the file to read.
     */
    private static ByteBuffer fileToByteBuffer(String path) throws java.io.FileNotFoundException, java.io.IOException {
        InputStream inputStream = new FileInputStream(path);
        byte[] bytes = inputStream.readAllBytes();
        inputStream.close();
        return ByteBuffer.wrap(bytes);
    }
}